modules/gapi/samples/face_beautification.cpp

```diff
@@ -0,0 +1,797 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level
+// directory of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2018-2019 Intel Corporation
+
+#if defined(HAVE_OPENCV_GAPI) && defined(HAVE_INF_ENGINE)
```

Comment:
> `HAVE_OPENCV_GAPI`

Like other OpenCV modules it can be tested only after including of `<opencv2/opencv_modules.hpp>`.
Anyway, it is redundant and should be removed - this is G-API sample (located in `gapi/samples`).

---

> `HAVE_INF_ENGINE`

This check is much worse. There is no header with this definition (`cvconfig.h` has some build options, but this header is private and it is not allowed in samples).

And even these definitions would not help to build sample with IE, because IE is not in dependency list of this sample.

This definition should be passed through CMake scripts.
After `ocv_add_samples()` you need to add IE dependency:
```.diff
 ocv_add_samples()
+if(TARGET example_gapi_face_beautification AND INF_ENGINE_TARGET)
+  target_link_libraries(example_gapi_face_beautification PRIVATE ${INF_ENGINE_TARGET})
+endif()
```

See these [build logs](http://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/2879/steps/compile%20release/logs/stdio) of "Custom" builder (find "face_beautification" and check its compiler options)

Reply:
@alalek in fact `HAVE_INF_ENGINE` is not even required since G-API builds its IA API part even when there's no IE present

Reply:
Right, IE is not used by sample directly (IE headers are not used, linker dependency is not needed).

But some IE-related symbols are missing in G-API:
```
undefined reference to `cv::gapi::ie::backend()'
```
Consider adding fallback implementation of `cv::gapi::ie::backend()` with something like:
```
CV_Error(Error::StsNotImplemented, "This OpenCV build doesn't include InferenceEngine support");
```

and then remove this check from sample (make it runtime by capturing `StsNotImplemented` if needed).

Reply:
@dmatveev As I understand, you meant changes from your PR #15753, specifically commits [1](https://github.com/opencv/opencv/pull/15753/commits/655663620b2a18a09d678cb43023d404f4099a2c) and [2](https://github.com/opencv/opencv/pull/15753/commits/451d21a53f2c1869ed316cce270fef82c5d3281a); would you mind if I use them?

Reply:
Yes, I thought that was merged already, but it you're all right the problem is still here. Bless you to steal these changes into your PR.

Reply:
Thank you!